Implicit key type conversion for join using step in pipeline

vdimir opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Supports implicit key type conversion for JOIN. Closes #18567.

Ref PR #18672

vdimir committed:
Add converting step for 'join using'

vdimir committed:
Fix converting right key type in join using

vdimir committed:
Fix TableJoin, upd 01675_join_implicit_cast

akuzm commented on src/Interpreters/join_common.cpp:
```diff
@@ -283,6 +285,87 @@ void addDefaultValues(IColumn & column, const DataTypePtr & type, size_t count)
+{
+    assert(left_keys.size() == right_keys.size());
+
+    /// only JOIN USING supported
```
The fact that only `USING` is supported is clear enough from the code, the comment should probably tell "why".

vdimir replied:
I've added comment to TreeRewriter, where we start to deal with join column conversion and to InterpreterSelectQuery where step added to query plan.
https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-03665017349505c5529e5d0eeeeae80081e064267d6035b94cd428baf11856ceR417
https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-5452971ae18972aa669265008a29b3732a99d9415f5b3af10d573c0807667a22R978

akuzm commented on src/Interpreters/join_common.cpp:
```diff
@@ -283,6 +285,87 @@ void addDefaultValues(IColumn & column, const DataTypePtr & type, size_t count)
+        if (visited_left.contains(left_keys[i]) || visited_right.contains(right_keys[i]))
+        {
+            /// if one column joined with multiple different others do not perform conversion
+            /// e.g. `JOIN ... ON t1.a == t2.a AND t1.a == t2.b`
```
We only support `USING`, is this possible?

vdimir replied:
I've removed this code

akuzm commented on src/Interpreters/TableJoin.cpp:
```diff
@@ -233,9 +233,15 @@ void TableJoin::addJoinedColumnsAndCorrectNullability(ColumnsWithTypeAndName & c
         if (leftBecomeNullable(col.type))
             col.type = makeNullable(col.type);
 
+    std::unordered_map<String, DataTypePtr> type_map;
```
So the type information in the input block here is "wrong" (before conversion), and we have to overwrite it w/ the converted values? But why? Can the input block be made to already have the proper types?

KochetovNicolai replied:
I guess, if actions for converting types will be in a separate step, this code won't be needed ...

vdimir replied:
I moved  common type calculation into [TreeRewriter](https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-03665017349505c5529e5d0eeeeae80081e064267d6035b94cd428baf11856ceR421) and now TableJoin contains information about converted types on earlier stage, so it is possible to change types in [addJoinedColumn](https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-a8a50243ee6c63a7030a0f87fe3e35677f6097063a807ac0e6f4fd0f5dd3b77eR222) and have correct types in columns_added_by_join

akuzm reviewed (approved):
It's nice that this patch is much smaller than the previous version.

I think it would be helpful to add a high-level comment somewhere, that describes what we do -- e.g. we can convert `USING` columns to their common supertype by inserting additional casts into ExpressionTransform before join. This only works for `USING` because etc etc...

KochetovNicolai commented on src/Interpreters/ExpressionAnalyzer.cpp:
```diff
@@ -712,23 +711,33 @@ ArrayJoinActionPtr SelectQueryExpressionAnalyzer::appendArrayJoin(ExpressionActi
 {
-    JoinPtr table_join = makeTableJoin(*syntax->ast_join);
+    JoinCommon::JoinConvertActions converting_actions;
+    JoinPtr table_join = makeTableJoin(*syntax->ast_join, sample_block, converting_actions);
```
I think, instead of passing `sample_block` here, we can use `chain.getLastStep().getResultColumns()`. Cause we have added actions to chain in `appendJoinLeftKeys`

vdimir replied:
Yes, it works https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-885608312a315dd9c5ce4577294b073f1be45b4fb55eba485304861cf8e3e1edR743

KochetovNicolai commented on src/Interpreters/ExpressionAnalyzer.cpp:
```diff
@@ -712,23 +711,33 @@ ArrayJoinActionPtr SelectQueryExpressionAnalyzer::appendArrayJoin(ExpressionActi
-    ExpressionActionsChain::Step & step = chain.lastStep(columns_after_array_join);
+        before_join_dag = ActionsDAG::merge(std::move(*before_join_dag->clone()), std::move(*converting_actions.left_actions->clone()));
 
-    chain.steps.push_back(std::make_unique<ExpressionActionsChain::JoinStep>(
-            syntax->analyzed_join, table_join, step.getResultColumns()));
+        chain.steps.push_back(std::make_unique<ExpressionActionsChain::ExpressionActionsStep>(converting_actions.left_actions));
```
Here we change actions of previous step **and** add a new step to chain (which is not used anyway).
Changing of `before_join_dag` is not good here, because those actions are already in chain. It means, that `getResultColumns` for the last step will be changed - which may be unusual.

The correct thing to do is:
* add `converting_actions.left_actions` to chain (as it is already done. before_join_dag shouldn't be changed)
* return pair `(converting_actions.left_actions, JoinPtr)` from `appendJoin`
* add separate step to `ExpressionAnalysisResult` (something like `converting_join_types` between `before_join` and `join`
* add another one `ExpressionStep` before adding join (probably here: https://github.com/ClickHouse/ClickHouse/blob/a2ceca3fa18f1df125de02c7c09e3999f5df8fce/src/Interpreters/InterpreterSelectQuery.cpp#L978-L979)

vdimir replied:
Yes, it's really simplify code. Now actions for converting stored in [converting_join_columns](https://github.com/ClickHouse/ClickHouse/pull/19885/files#diff-5452971ae18972aa669265008a29b3732a99d9415f5b3af10d573c0807667a22R981)

KochetovNicolai commented on src/Interpreters/ExpressionAnalyzer.cpp:
```diff
@@ -1423,10 +1444,10 @@ ExpressionAnalysisResult::ExpressionAnalysisResult(
         {
-            query_analyzer.appendJoinLeftKeys(chain, only_types || !first_stage);
-
+            Block left_block_sample = source_header;
```
Note that `source_header` may already be not correct here, e.g. in case if query has `ARRAY JOIN`.
There should be used `ExpressionAnalyzerData::columns_after_array_join`.
But as I commented before, it's better to use `chain.getLastStep().getResultColumns()` inside `query_analyzer.appendJoinLeftKeys`

vdimir replied:
`chain.getLastStep().getResultColumns` helped

KochetovNicolai reviewed (changes requested):
.

vdimir committed:
Calculate common type for join using in TreeRewriter

Split before_join and converting_join_columns dags

Add more detailed comments for column type conversion for join using

vdimir commented:
I've found bugs with aggregation due to fact that in [analyzeAggregation](https://github.com/ClickHouse/ClickHouse/blob/229acbad5d9bb7e7a213b190bd9d6fd4aec18de4/src/Interpreters/ExpressionAnalyzer.cpp#L216) corrected types are not calculated yet. Common type calculated moved to earlier stage to `TreeRewriter` and type mapping stored in `TableJoin`. It fix bug and also I suppose makes code a bit clearer

vdimir commented on src/Interpreters/SubqueryForSet.cpp:
```diff
@@ -39,10 +39,20 @@ void SubqueryForSet::renameColumns(Block & block)
+    else
+    {
+        auto new_dag = ActionsDAG::merge(
+            std::move(*joined_block_actions->getActionsDAG().clone()),
```
Should we split `joined_block_actions` into separate variables not to use `merge` here? Now it consist from [createJoinedBlockActions](https://github.com/ClickHouse/ClickHouse/blob/a6ea5ff682a2233f9d7983b4459012ec01ba1cb2/src/Interpreters/ExpressionAnalyzer.cpp#L763) and additional converting step. @KochetovNicolai

KochetovNicolai replied:
As I understand, this actions are applied to right part of join. And we don't use separate step for this actions for now. So, I think it is ok to merge.

vdimir committed:
Give up on name mismatch in inferJoinKeyCommonType

vdimir committed:
Fix style in appendJoin

vdimir committed:
Add tests to 01720_join_implicit_cast - switch, use_nulls

vdimir committed:
Perform implicit type conversion for JOIN ON keys

vdimir committed:
Remove addRequiredLeftColumn, some tests for join on different types

vdimir committed:
Fix converting join on keys, move actions into TableJoin

vdimir committed:
Add comments to TableJoin::inferJoinKeyCommonType

vdimir committed:
More tests for join on key type convert

vdimir committed:
Remove 'error' from comment from 01710_join_use_nulls

vdimir committed:
Split test join_implicit_cast (due to timeouts in flaky checks)

vdimir commented:
I've added implementation of type conversion for `JOIN ON` syntax. It don't change types in result table. For that conversion do not replace original columns, but and new ones and join performed by this new columns.

vdimir committed:
Add tests to join_implicit_cast checks column name clash

KochetovNicolai reviewed (approved):
I did not read tests, but changes in code LGTM.

vdimir committed:
Merge remote-tracking branch 'upstream/master' into join-cast-types-v2

vdimir committed:
Merge branch 'master' into join-cast-types-v2

akuzm commented:
Integration test failure doesn't look relevant.
MSan fuzzer looks like this one https://github.com/ClickHouse/ClickHouse/issues/20046

Should we merge?

vdimir commented:
> Should we merge?

I found one issue, query doesn't work if syntax `JOIN ON` used with some constant in `WHERE`, e.g. `SELECT * FROM t1 FULL JOIN t2 ON t1.a = t2.a WHERE 1`. It shows error `DB::Exception: Not found column t2.a in block. There are only columns: CAST(t2.a, Int32)`. If type conversion for `t1.a` and `t2.a` not needed it works ok as earlier. 

Likely after some optimization some required columns are dropped. I'm trying to figure it out. 
Maybe we can merge this and fix this issue later or it's better to investigate it more detailed?

akuzm commented:
> It shows error `DB::Exception: Not found column t2.a in block. There are only columns: CAST(t2.a, Int32)`.

Looks kinda bad, probably better to investigate. Do I remember correctly that you keep the name of the column, but replace it with the casted one? So it should have been named just `t2.a` and not `cast...` if everything worked correctly.

vdimir committed:
Refactor join: make IJoin implementations independent from TableJoin

vdimir committed:
Revert "Refactor join: make IJoin implementations independent from TableJoin"

This reverts commit e6406c3f4c9e53c371bbc9f79cee8a870dec58e0.

vdimir committed:
Clean converting-related fields in TableJoin::resetCollected

vdimir committed:
Clean subqueries_for_sets before second analyze in InterpreterSelectQuery

vdimir committed:
Add test for cast join on keys with second analyze

vdimir committed:
Add logging for need_analyze_again in InterpreterSelectQuery

vdimir committed:
Fixes for join key inferring

Fix error with wrong right column type in aggregate with nulls

Add logging inferred type for join keys

Add test for join_use_nulls

olgarev commented:
Internal documentation tickey DOCSUP-9868
